File: cpp/sorting/bubble_sort.cpp
#include <iostream>
#include <vector>
using namespace std;

void bubbleSort(vector<int> &arr) {
    int n = arr.size();
    for(int i=0; i<n; i++){
        bool swapped = false;
        for(int j=0; j<n-i-1; j++){
            if(arr[j] > arr[j+1]){
                swap(arr[j], arr[j+1]);
                swapped = true;
            }
        }
        if(!swapped) break;
    }
}

void printArray(const vector<int> &arr){
    for(int num: arr) cout << num << " ";
    cout << endl;
}

int main(){
    vector<int> arr = {64, 34, 25, 12, 22, 11, 90};
    cout << "Original: ";
    printArray(arr);

    bubbleSort(arr);
    cout << "Bubble Sorted: ";
    printArray(arr);

    return 0;
}

File: python/sorting/bubble_sort.py
def bubble_sort_optimized(arr):
    n = len(arr)

    for i in range(n):
        swapped = False

        for j in range(0, n - i - 1):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                swapped = True

        # If no two elements were swapped, the list is sorted
        if not swapped:
            break

    return arr


# Example usage
data = [5, 6 , 1 , 3, 4]
print(bubble_sort_optimized(data))
